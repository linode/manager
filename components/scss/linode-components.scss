@import 'mixins';
@import 'variables';
@import 'bootstrap';
@import 'utilities';

@import 'components/index';

html {
  font-size: 62.5%; // resets root font size to 10px
}

// default font sizes, defined in rem with a pixel fallback, computed whole pixel values
.font-normal {
  @include font-size(1.5);
}
.font-small {
  @include font-size(1.3);
}
.font-xs {
  @include font-size(1.2);
}
.font-medium {
  @include font-size(1.6);
}
.font-large {
  @include font-size(1.7);
}
.font-extra-large {
  @include font-size(1.8);
}

body {
  @extend .font-normal;
  background: $background-color;
  color: $black;
  font-weight: normal;
}

h1, h2, h3, h4, h5, h6 {
  margin: 0;
  font-weight: normal;
  line-height: normal;
  color: $black;
}

h1 {
  @include font-size(2.8);
  margin-bottom: 15px;
}

h2 {
  @include font-size(2.2);
}

h3 {
  @extend .font-large;
}

h4 {
  @extend .font-medium;
}

h5, h6 {
  @extend .font-normal;
}

a, a.force-link:not([href]):not([tabindex]) {
    color: $blue;
    text-decoration: none;
    cursor: pointer;

    &:hover, &:focus, &:active {
        text-decoration: underline;
    }
}

a > i.fa {
  color: $blue;
}

label {
  margin-bottom: 0;
}

p {
  margin-bottom: $main-padding-tb/2;
}

hr {
    width: 100%;
    margin: 0;
    background: $gray;
}

small {
  @extend .font-small;
}

strong {
  font-weight: 500; // toned down from 'bolder'
}

section {
    margin-bottom: 15px;
}

.text-muted {
    // Bootstrap sets !important here by default;
    color: $lightest-black !important;
}

.full-height {
    height: 100%;
}

.navicon {
    @extend .font-large;
    padding-left: 15px;
    position: relative;
    top: 2px;

    &:hover,
    &:active,
    &:focus {
        color: $blue;
    }
}

footer {
    @extend .font-small;

    padding: 15px 0;

    span, a {
        color: $lightest-black;
        padding: 0 10px;
    }

    a {
        text-decoration: underline;
    }
}

.Error {
    text-align: center;

    h1 {
        font-size: 8rem; /* Special font-size case */
        color: $dark-gray;
    }

    &-body {
        margin-top: 10px;
    }
}

.sub-header {
    color: $light-black;
    padding-top: 15px;
    padding-bottom: 5px;
}

header .sub-header {
    padding-bottom: 0px;
}

.list-unstyled {
    margin-bottom: 0;
}

.row-justify {
    @extend .row;
    display: flex;
    flex-wrap: wrap;

    margin-left: -$grid-gutter-width-base / 4;
    margin-right: -$grid-gutter-width-base / 4;

    & > [class^="col-"] {
        padding-left: $grid-gutter-width-base / 4;
        padding-right: $grid-gutter-width-base / 4;
    }
}

a:hover, a:focus, a:active {
    color: $blue;
}

.alert {
    @extend .font-xs;

    display: inline-block;
    margin: 5px 0 0 0;
    padding: 15px;
    line-height: initial;
    min-height: initial;
}
