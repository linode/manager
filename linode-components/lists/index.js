export { default as List } from './List';
export { default as ScrollingList } from './ScrollingList';
