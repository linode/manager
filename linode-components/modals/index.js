export { default as ModalShell } from './ModalShell';
export { default as ConfirmModalBody } from './ConfirmModalBody';
export { default as DeleteModalBody } from './DeleteModalBody';
