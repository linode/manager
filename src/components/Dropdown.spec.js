import React from 'react';
import sinon from 'sinon';
import { shallow, mount } from 'enzyme';
import { Dropdown } from 'linode-components';
import { Button } from 'linode-components';

describe('components/Dropdown', () => {
  const sandbox = sinon.sandbox.create();

  afterEach(() => {
    sandbox.restore();
  });

  it('renders dropdown component 2', () => {
    const dropdown = mount(
      <Dropdown
        groups={[
          { elements: [{ action: () => { }, name: 'Drew' }] },
          {
            elements: [
              { action: () => { }, name: 'Phil' },
              { action: () => { }, name: 'Will' },
            ],
          },
        ]}
      />
    );

    // Should find two buttons.
    expect(dropdown.findWhere((n) => n.type() === Button && n.text() === 'Drew').length).toBe(1);

    expect(dropdown.find('.Dropdown-item').length).toBe(3);
    expect(dropdown.find('.Dropdown-item').at(1).text()).toBe('Phil');
    expect(dropdown.find('.Dropdown-item').at(2).text()).toBe('Will');
  });

  it('clickable dropdown', () => {
    const clickFirst = sandbox.spy();
    const clickBodyItem = sandbox.spy();
    const unclicked = sandbox.spy();
    const dropdown = shallow(
      <Dropdown
        groups={[
          { elements: [{ action: clickFirst, name: 'Drew' }] },
          {
            elements: [
              { action: clickBodyItem, name: 'Phil' },
              { action: unclicked, name: 'Will' },
            ],
          },
        ]}
      />
    );

    // Click first item should trigger clickFirst
    dropdown.find('.Dropdown-first').simulate('click');
    // Click toggle should open body
    dropdown.find('.Dropdown-toggle').simulate('click');
    // Click first menu item should trigger clickBodyItem
    dropdown.find('.Dropdown-item').at(1).simulate('mousedown');
    // Last menu item goes unclicked

    expect(clickFirst.callCount).toBe(1);
    expect(dropdown.find('.Dropdown--open').length).toBe(1);
    expect(clickBodyItem.callCount).toBe(1);
    expect(unclicked.callCount).toBe(0);

    // Mousedown did not hide dropdown
    expect(dropdown.find('.Dropdown--open').length).toBe(1);
    // But onblur does
    dropdown.find('.Dropdown').simulate('blur');
    expect(dropdown.find('.Dropdown--open').length).toBe(0);
  });

  it('closes on second click', () => {
    const noAction = () => { };
    const dropdown = shallow(
      <Dropdown
        groups={[
          { elements: [{ action: noAction, name: '' }] },
          { elements: [{ action: noAction, name: '' }] },
        ]}
      />
    );

    dropdown.find('.Dropdown-toggle').simulate('click');
    expect(dropdown.find('.Dropdown--open').length).toBe(1);
    dropdown.find('.Dropdown-toggle').simulate('click');
    expect(dropdown.find('.Dropdown--open').length).toBe(0);
  });

  it('closes on blur', () => {
    const noAction = () => { };
    const dropdown = shallow(
      <Dropdown
        groups={[
          { elements: [{ action: noAction, name: '' }] },
          { elements: [{ action: noAction, name: '' }] },
        ]}
      />
    );

    const toggle = dropdown.findWhere((n) => {
      return n.type() === Button && n.hasClass('Dropdown-toggle');
    });

    toggle.simulate('click');
    expect(dropdown.find('.Dropdown--open').length).toBe(1);

    dropdown.simulate('blur');
    expect(dropdown.find('.Dropdown--open').length).toBe(0);
  });
});
