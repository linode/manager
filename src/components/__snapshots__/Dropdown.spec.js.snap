// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Dropdown should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Dropdown
    analytics={Object {}}
    dropdownIcon="fa-caret-down"
    duplicateFirst={true}
    groups={
        Array [
            Object {
              "elements": Array [
                Object {
                  "action": [Function],
                  "name": "Drew",
                },
              ],
            },
            Object {
              "elements": Array [
                Object {
                  "action": [Function],
                  "name": "Phil",
                },
                Object {
                  "action": [Function],
                  "name": "Will",
                },
              ],
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Button
          buttonClass="btn-default"
          className="Dropdown-first"
          disabled={false}
          id="drew"
          onClick={[Function]}
          to={undefined}
          type="button"
>
          Drew
</Button>,
        <Button
          aria-expanded={false}
          aria-haspopup="true"
          buttonClass="btn-default"
          className="Dropdown-toggle"
          data-toggle="dropdown"
          disabled={false}
          id=""
          onClick={[Function]}
          type="button"
>
          <i
                    className="fa fa-caret-down"
          />
</Button>,
        <div
          className="Dropdown-menu Dropdown-menu--right"
>
          <div
                    className="Dropdown-group"
          >
                    <div
                              className="Dropdown-elements"
                    >
                              <button
                                        className="Dropdown-item"
                                        id="drew"
                                        onMouseDown={[Function]}
                                        type="button"
                              >
                                        Drew
                              </button>
                    </div>
          </div>
          <div
                    className="Dropdown-group"
          >
                    <div
                              className="Dropdown-elements"
                    >
                              <button
                                        className="Dropdown-item"
                                        id="phil"
                                        onMouseDown={[Function]}
                                        type="button"
                              >
                                        Phil
                              </button>
                              <button
                                        className="Dropdown-item"
                                        id="will"
                                        onMouseDown={[Function]}
                                        type="button"
                              >
                                        Will
                              </button>
                    </div>
          </div>
</div>,
      ],
      "className": "Dropdown btn-group ",
      "onBlur": [Function],
      "tabIndex": "-1",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "buttonClass": "btn-default",
          "children": Array [
            null,
            "Drew",
          ],
          "className": "Dropdown-first",
          "disabled": false,
          "id": "drew",
          "onClick": [Function],
          "to": undefined,
          "type": "button",
        },
        "ref": null,
        "rendered": Array [
          null,
          "Drew",
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "aria-expanded": false,
          "aria-haspopup": "true",
          "buttonClass": "btn-default",
          "children": <i
            className="fa fa-caret-down"
/>,
          "className": "Dropdown-toggle",
          "data-toggle": "dropdown",
          "disabled": false,
          "id": "",
          "onClick": [Function],
          "type": "button",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fa fa-caret-down",
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="Dropdown-group"
>
              <div
                            className="Dropdown-elements"
              >
                            <button
                                          className="Dropdown-item"
                                          id="drew"
                                          onMouseDown={[Function]}
                                          type="button"
                            >
                                          Drew
                            </button>
              </div>
</div>,
            <div
              className="Dropdown-group"
>
              <div
                            className="Dropdown-elements"
              >
                            <button
                                          className="Dropdown-item"
                                          id="phil"
                                          onMouseDown={[Function]}
                                          type="button"
                            >
                                          Phil
                            </button>
                            <button
                                          className="Dropdown-item"
                                          id="will"
                                          onMouseDown={[Function]}
                                          type="button"
                            >
                                          Will
                            </button>
              </div>
</div>,
          ],
          "className": "Dropdown-menu Dropdown-menu--right",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                null,
                <div
                  className="Dropdown-elements"
>
                  <button
                                    className="Dropdown-item"
                                    id="drew"
                                    onMouseDown={[Function]}
                                    type="button"
                  >
                                    Drew
                  </button>
</div>,
              ],
              "className": "Dropdown-group",
            },
            "ref": null,
            "rendered": Array [
              null,
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <button
                      className="Dropdown-item"
                      id="drew"
                      onMouseDown={[Function]}
                      type="button"
>
                      Drew
</button>,
                  ],
                  "className": "Dropdown-elements",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": "Drew",
                    "nodeType": "host",
                    "props": Object {
                      "children": "Drew",
                      "className": "Dropdown-item",
                      "id": "drew",
                      "onMouseDown": [Function],
                      "type": "button",
                    },
                    "ref": null,
                    "rendered": "Drew",
                    "type": "button",
                  },
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": Array [
                null,
                <div
                  className="Dropdown-elements"
>
                  <button
                                    className="Dropdown-item"
                                    id="phil"
                                    onMouseDown={[Function]}
                                    type="button"
                  >
                                    Phil
                  </button>
                  <button
                                    className="Dropdown-item"
                                    id="will"
                                    onMouseDown={[Function]}
                                    type="button"
                  >
                                    Will
                  </button>
</div>,
              ],
              "className": "Dropdown-group",
            },
            "ref": null,
            "rendered": Array [
              null,
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <button
                      className="Dropdown-item"
                      id="phil"
                      onMouseDown={[Function]}
                      type="button"
>
                      Phil
</button>,
                    <button
                      className="Dropdown-item"
                      id="will"
                      onMouseDown={[Function]}
                      type="button"
>
                      Will
</button>,
                  ],
                  "className": "Dropdown-elements",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": "Phil",
                    "nodeType": "host",
                    "props": Object {
                      "children": "Phil",
                      "className": "Dropdown-item",
                      "id": "phil",
                      "onMouseDown": [Function],
                      "type": "button",
                    },
                    "ref": null,
                    "rendered": "Phil",
                    "type": "button",
                  },
                  Object {
                    "instance": null,
                    "key": "Will",
                    "nodeType": "host",
                    "props": Object {
                      "children": "Will",
                      "className": "Dropdown-item",
                      "id": "will",
                      "onMouseDown": [Function],
                      "type": "button",
                    },
                    "ref": null,
                    "rendered": "Will",
                    "type": "button",
                  },
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Button
            buttonClass="btn-default"
            className="Dropdown-first"
            disabled={false}
            id="drew"
            onClick={[Function]}
            to={undefined}
            type="button"
>
            Drew
</Button>,
          <Button
            aria-expanded={false}
            aria-haspopup="true"
            buttonClass="btn-default"
            className="Dropdown-toggle"
            data-toggle="dropdown"
            disabled={false}
            id=""
            onClick={[Function]}
            type="button"
>
            <i
                        className="fa fa-caret-down"
            />
</Button>,
          <div
            className="Dropdown-menu Dropdown-menu--right"
>
            <div
                        className="Dropdown-group"
            >
                        <div
                                    className="Dropdown-elements"
                        >
                                    <button
                                                className="Dropdown-item"
                                                id="drew"
                                                onMouseDown={[Function]}
                                                type="button"
                                    >
                                                Drew
                                    </button>
                        </div>
            </div>
            <div
                        className="Dropdown-group"
            >
                        <div
                                    className="Dropdown-elements"
                        >
                                    <button
                                                className="Dropdown-item"
                                                id="phil"
                                                onMouseDown={[Function]}
                                                type="button"
                                    >
                                                Phil
                                    </button>
                                    <button
                                                className="Dropdown-item"
                                                id="will"
                                                onMouseDown={[Function]}
                                                type="button"
                                    >
                                                Will
                                    </button>
                        </div>
            </div>
</div>,
        ],
        "className": "Dropdown btn-group ",
        "onBlur": [Function],
        "tabIndex": "-1",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "buttonClass": "btn-default",
            "children": Array [
              null,
              "Drew",
            ],
            "className": "Dropdown-first",
            "disabled": false,
            "id": "drew",
            "onClick": [Function],
            "to": undefined,
            "type": "button",
          },
          "ref": null,
          "rendered": Array [
            null,
            "Drew",
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "aria-expanded": false,
            "aria-haspopup": "true",
            "buttonClass": "btn-default",
            "children": <i
              className="fa fa-caret-down"
/>,
            "className": "Dropdown-toggle",
            "data-toggle": "dropdown",
            "disabled": false,
            "id": "",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "fa fa-caret-down",
            },
            "ref": null,
            "rendered": null,
            "type": "i",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="Dropdown-group"
>
                <div
                                className="Dropdown-elements"
                >
                                <button
                                                className="Dropdown-item"
                                                id="drew"
                                                onMouseDown={[Function]}
                                                type="button"
                                >
                                                Drew
                                </button>
                </div>
</div>,
              <div
                className="Dropdown-group"
>
                <div
                                className="Dropdown-elements"
                >
                                <button
                                                className="Dropdown-item"
                                                id="phil"
                                                onMouseDown={[Function]}
                                                type="button"
                                >
                                                Phil
                                </button>
                                <button
                                                className="Dropdown-item"
                                                id="will"
                                                onMouseDown={[Function]}
                                                type="button"
                                >
                                                Will
                                </button>
                </div>
</div>,
            ],
            "className": "Dropdown-menu Dropdown-menu--right",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  null,
                  <div
                    className="Dropdown-elements"
>
                    <button
                                        className="Dropdown-item"
                                        id="drew"
                                        onMouseDown={[Function]}
                                        type="button"
                    >
                                        Drew
                    </button>
</div>,
                ],
                "className": "Dropdown-group",
              },
              "ref": null,
              "rendered": Array [
                null,
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <button
                        className="Dropdown-item"
                        id="drew"
                        onMouseDown={[Function]}
                        type="button"
>
                        Drew
</button>,
                    ],
                    "className": "Dropdown-elements",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": "Drew",
                      "nodeType": "host",
                      "props": Object {
                        "children": "Drew",
                        "className": "Dropdown-item",
                        "id": "drew",
                        "onMouseDown": [Function],
                        "type": "button",
                      },
                      "ref": null,
                      "rendered": "Drew",
                      "type": "button",
                    },
                  ],
                  "type": "div",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  null,
                  <div
                    className="Dropdown-elements"
>
                    <button
                                        className="Dropdown-item"
                                        id="phil"
                                        onMouseDown={[Function]}
                                        type="button"
                    >
                                        Phil
                    </button>
                    <button
                                        className="Dropdown-item"
                                        id="will"
                                        onMouseDown={[Function]}
                                        type="button"
                    >
                                        Will
                    </button>
</div>,
                ],
                "className": "Dropdown-group",
              },
              "ref": null,
              "rendered": Array [
                null,
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <button
                        className="Dropdown-item"
                        id="phil"
                        onMouseDown={[Function]}
                        type="button"
>
                        Phil
</button>,
                      <button
                        className="Dropdown-item"
                        id="will"
                        onMouseDown={[Function]}
                        type="button"
>
                        Will
</button>,
                    ],
                    "className": "Dropdown-elements",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": "Phil",
                      "nodeType": "host",
                      "props": Object {
                        "children": "Phil",
                        "className": "Dropdown-item",
                        "id": "phil",
                        "onMouseDown": [Function],
                        "type": "button",
                      },
                      "ref": null,
                      "rendered": "Phil",
                      "type": "button",
                    },
                    Object {
                      "instance": null,
                      "key": "Will",
                      "nodeType": "host",
                      "props": Object {
                        "children": "Will",
                        "className": "Dropdown-item",
                        "id": "will",
                        "onMouseDown": [Function],
                        "type": "button",
                      },
                      "ref": null,
                      "rendered": "Will",
                      "type": "button",
                    },
                  ],
                  "type": "div",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
